emily/handler/src/database/entries/deposit.rs

```diff
@@ -286,3 +325,95 @@ impl From<DepositInfoEntry> for DepositInfo {
         }
     }
 }
+
+/// Validated version of the update deposit request.
+pub struct ValidatedUpdateDepositsRequest {
+    /// Validated deposit update requests.
+    pub deposits: Vec<ValidatedDepositUpdate>,
+}
+
+/// Implement try from for the validated depoit requests.
+impl TryFrom<UpdateDepositsRequestBody> for ValidatedUpdateDepositsRequest {
+    type Error = Error;
+    fn try_from(update_request: UpdateDepositsRequestBody) -> Result<Self, Self::Error> {
+        // Validate all the depoit updates.
+        let deposits = update_request
+            .deposits
+            .into_iter()
+            .map(|i| i.try_into())
+            .collect::<Result<_, Error>>()?;
+        Ok(ValidatedUpdateDepositsRequest { deposits })
+    }
+}
+
+/// Validated deposit update.
+pub struct ValidatedDepositUpdate {
+    /// Key.
+    pub key: DepositEntryKey,
+    /// Deposit event.
+    pub event: DepositEvent,
+}
+
+impl TryFrom<DepositUpdate> for ValidatedDepositUpdate {
+    type Error = Error;
+    fn try_from(update: DepositUpdate) -> Result<Self, Self::Error> {
+        // Make key.
+        let key = DepositEntryKey {
+            bitcoin_tx_output_index: update.bitcoin_tx_output_index,
+            bitcoin_txid: update.bitcoin_txid,
+        };
+        // Make status entry.
+        let status_entry: StatusEntry = match update.status {
+            Status::Accepted => {
+                let fulfillment = update.fulfillment.ok_or(Error::InternalServer)?;
+                StatusEntry::Accepted(fulfillment)
+            }
+            Status::Confirmed => StatusEntry::Confirmed,
+            Status::Pending => StatusEntry::Pending,
+            Status::Reevaluating => StatusEntry::Reevaluating,
+            Status::Failed => StatusEntry::Failed,
+        };
+        // Make the new event.
+        let event = DepositEvent {
+            status: status_entry,
+            message: update.status_message,
+            stacks_block_height: update.last_update_height,
+            stacks_block_hash: update.last_update_block_hash,
+        };
+        // Return the validated update.
+        Ok(ValidatedDepositUpdate { key, event })
+    }
+}
+
+/// Packaged deposit update.
+pub struct DepositUpdatePackage {
+    /// Key.
+    pub key: DepositEntryKey,
+    /// Version.
+    pub version: u64,
+    /// Deposit event.
+    pub event: DepositEvent,
+}
+
+/// Implementation of deposit update package.
+impl DepositUpdatePackage {
+    /// Implements from.
+    pub fn from(entry: &DepositEntry, update: ValidatedDepositUpdate) -> Result<Self, Error> {
+        // Ensure the keys are equal.
+        if update.key != entry.key {
+            return Err(Error::Debug(
+                "Attempted to update deposit txid + output index combo".into(),
+            ));
+        }
+        // Ensure that this event is valid if it follows the current latest event.
+        entry
+            .latest_event()
```

Comment:
Don't we need to validate the entry first?

Reply:
Made this return a result instead!